ASoC: rt712-sdca: add the function for version B

shumingfan opened the pull request:
The version B will support the multi-lane function and integrate the DMIC function in one SoundWire interface.
Due to some registers having different default values between version A and B, this patch also removes the redundant default registers to avoid confusion.

plbossart commented on sound/soc/codecs/rt712-sdca.c:
```diff
@@ -763,21 +771,21 @@ static int rt712_sdca_fu05_event(struct snd_soc_dapm_widget *w,
 		regmap_write(rt712->regmap,
 			SDW_SDCA_CTL(FUNC_NUM_JACK_CODEC, RT712_SDCA_ENT_USER_FU05,
-				RT712_SDCA_CTL_FU_MUTE, CH_R),
+				RT712_SDCA_CTL_FU_MUTE, CH_02),
```
it might be simpler to review is the CH_L/CH_01 and CH_R/CH_02 were moved to an initial patch without any change of functionality?

shumingfan replied:
Thanks for reviewing. I will make an initial patch for this change.

plbossart commented on sound/soc/codecs/rt712-sdca.h:
```diff
@@ -207,6 +238,11 @@ enum rt712_sdca_hw_id {
+	RT712_VA,
+	RT712_VB,
+};
+
```
is there anything that differentiates the VA and VB at the device_ID level?
Also is there a way to know if the dmic is supported or not based on ACPI information?
not sure how we would change the machine driver to add the dmic path or not?

@bardliao what do you think?

bardliao replied:
We could use `dai_info->quirk` to make the DMIC be optional, like https://github.com/thesofproject/linux/pull/4848/commits/47bae326da237ca6cf0414dc5a8a823d1ff83a39

plbossart replied:
yes but that quirk is defined at the machine driver level. It's too late to select the topology, which is entirely based on ACPI information.

plbossart commented on sound/soc/codecs/rt712-sdca.h:
```diff
@@ -191,6 +221,7 @@ struct  rt712_sdca_priv {
 enum {
 	RT712_AIF1,
 	RT712_AIF2,
+	RT712_AIF3,
```
how can we make sure that AIF3 is never used with a VA chip?

I find the whole thing a bit complicated if the information is hidden the machine integration could be problematic, no?

also how would we select a topology that sometimes has a dmic dailink and sometimes not?

bardliao replied:
It is better to have different ADR between -VA and -VB. Or we might construct topology name dynamically, something like https://github.com/thesofproject/linux/pull/4847/commits/0930b9424667422db3fc7695bb525e89fa530e9d. But to be honest, I think it will be painful to do it.

plbossart reviewed (commented):
Thanks for sharing @shumingfan, I have rather big concerns on the topology/machine driver support if the VA/VB information is not visible at the ACPI level.

shumingfan committed:
ASoC: rt712-sdca: change the definition name of SDCA channel number

This patch doesn't have any change of functionality.

Signed-off-by: Shuming Fan <shumingf@realtek.com>

plbossart commented:
@shumingfan is the function listed in the DSDT? IIRC all the functions should be a "Device" under the codec device, so in theory we might be able to add a quirk to go read the function and decide which system it is.

IOW, we would first start from the ACPI _ADR, but add a quirk to go check if there is a subdevice representing the DMIC function. And in that case we would know for sure what the codec is and select topology.

shumingfan commented:
> IOW, we would first start from the ACPI _ADR, but add a quirk to go check if there is a subdevice representing the DMIC function. And in that case we would know for sure what the codec is and select topology.

@plbossart Regarding RT712VA, we might check ACPI _ADR that the part id 171X exists or not.
Regarding RT712VB or VA, could we check the DisCo property "mipi-sdca-terminal-type"?
If mipi-sdca-terminal-type belongs to 0x201 or 0x205, that machine will use the DMIC function.
![image](https://github.com/thesofproject/linux/assets/18300972/ff6f3230-61f0-4c4c-bdf3-7208555321ee)

plbossart commented:
> @plbossart Regarding RT712VA, we might check ACPI _ADR that the part id 171X exists or not. 

Yes that's a possibility.

>Regarding RT712VB or VA, could we check the DisCo property "mipi-sdca-terminal-type"? If mipi-sdca-terminal-type belongs to 0x201 or 0x205, that machine will use the DMIC function.

It's probably better to detect the function type, we should see a SmartMic function.

I'll start cooking something.

shumingfan commented:
> Of course that assumes ACPI information is valid, which is already not true on the first device I tried...

@plbossart  Thanks. The PR#4930 looks good to me. The function type of DMIC of RT713 might be a simple mic function.
![image](https://github.com/thesofproject/linux/assets/18300972/a6b2ffbe-8b07-4d1b-8751-4d538eed23b3)

shumingfan committed:
ASoC: rt712-sdca: add the function for version B

The version B will support the multi-lane function and integrate the DMIC function
in one SoundWire interface.
Due to some registers having different default values between version A and B,
this patch also removes the redundant default registers to avoid confusion.

Signed-off-by: Shuming Fan <shumingf@realtek.com>

plbossart commented on sound/soc/codecs/rt712-sdca-sdw.c:
```diff
@@ -178,13 +194,15 @@ static int rt712_sdca_read_prop(struct sdw_slave *slave)
 	struct sdw_dpn_prop *dpn;
 
+	sdw_slave_read_prop(slave);
+
```
how is this related to the VA/VB support?

shumingfan replied:
@plbossart This is related to the VB capability for multi-lanes. 
I tested with the PR https://github.com/thesofproject/linux/pull/4733.
I think there is no effect on the VA.

plbossart commented on sound/soc/codecs/rt712-sdca-sdw.c:
```diff
@@ -178,13 +194,15 @@ static int rt712_sdca_read_prop(struct sdw_slave *slave)
 
 	/* first we need to allocate memory for set bits in port lists */
-	prop->source_ports = BIT(4); /* BITMAP: 00010000 */
+	prop->source_ports = BIT(8) | BIT(4); /* BITMAP: 100010000 */
```
I am really confused, is the port8 supported on all versions?

shumingfan replied:
The port8 supports on the VB. 
The DMIC function will use port8 and the driver only creates the "rt712-sdca-aif3" of DMIC DAI when using the VB.
It is easy to add port8 here and allocate memory for set bits in port lists.
The VA could not use port8 because it would not create the DMIC DAI.

plbossart commented on sound/soc/codecs/rt712-sdca.c:
```diff
@@ -1031,6 +1432,10 @@ static int rt712_sdca_pcm_hw_params(struct snd_pcm_substream *substream,
 
+	/* VA doesn't support AIF3 */
+	if (dai->id == RT712_AIF3 && rt712->version_id == RT712_VA)
+		return -EINVAL;
```
in what cases would it be necessary to the AIF3?

I am really concerned that the machine driver would need to know about VA/VB.

It's really problematic if the same device has different capabilities over time. It's no longer the same device IMHO and the partID should have been modified.

shumingfan replied:
AIF3 is the DMIC function. The VB integrates the DMIC function in one soundwire interface which supports the multi-lanes. The customers could use the DMIC of RT712 if they don't want another DMIC codec.

> It's really problematic if the same device has different capabilities over time. It's no longer the same device IMHO and the partID should have been modified.

Hmm...I'm sorry for that. I can't decide the partID if the codec will be modified.

plbossart reviewed (changes requested):
@shumingfan I really don't know how we can support the VB at the machine driver level, this is really really problematic. The opens raised in the initial review have not been solved....

plbossart commented:
@shumingfan for my education, what's the difference between RT712 VB and RT722? They both seem to support UAJ, SmartAmp and SmartMic functions, along with multi-lane.

shumingfan commented:
> @shumingfan for my education, what's the difference between RT712 VB and RT722? They both seem to support UAJ, SmartAmp and SmartMic functions, along with multi-lane.

@plbossart As far as I know, RT712VB has a newer SDCA version than RT722.
RT712VB has another function like TV mode, however, it isn't implemented in the codec driver yet.
The basic functions like UAJ/Amp/Mic are similar to RT722.

plbossart commented on sound/soc/codecs/rt712-sdca.c:
```diff
@@ -1307,6 +1716,131 @@ int rt712_sdca_io_init(struct device *dev, struct sdw_slave *slave)
+
+	if (rt712->version_id == RT712_VA)
+		rt712_sdca_va_io_init(rt712);
+	else {
```
I would suggest adding a test here

````c
if (!sdca_device_quirk_match(slave, SDCA_QUIRKS_RT712_VB)) {
    dev_err(&slave->dev, "%s: RT712 VB detected with no SmartMic Function in ACPI, invalid configuration\n", __func__);
    return -EINVAL;
}
````

that way we will not have a case where we trip between hardware and ACPI config.

plbossart reviewed (commented):
@shumingfan I suggest rebasing on https://github.com/thesofproject/linux/pull/5010 to make sure ACPI and hardware are aligned.

shumingfan commented:
The patches were upstreamed already.
